petastorm/tests/test_pytorch_dataloader.py

```diff
@@ -248,65 +253,93 @@ def test_batched_data_loader_with_in_memory_cache(two_columns_non_petastorm_data
+            retrieved_so_far = None
             for idx in range(5):
                 batch = next(it)
-                # Assert that a new buffer is created inside the loader
-                assert loader._shuffling_buffer != first_buffer
+                if idx == 0:
+                    first_buffer = loader._shuffling_buffer
+
                 this_batch = batch['col_0'].clone()
                 assert list(this_batch.shape)[0] == batch_size
-                intersection = set(this_batch.tolist()).intersection(set(retrieved_in_first_epoch.tolist()))
-                assert len(intersection) == batch_size
-                retrieved_so_far = torch.cat([retrieved_so_far, this_batch], 0)
 
-        if num_epochs == 2:
-            with pytest.raises(StopIteration):
+                if retrieved_so_far is None:
+                    retrieved_so_far = this_batch
+                else:
+                    intersect = set(retrieved_so_far.tolist()).intersection(set(this_batch.tolist()))
+                    assert not intersect
+                    retrieved_so_far = torch.cat([retrieved_so_far, this_batch], 0)
+
+            retrieved_in_first_epoch = retrieved_so_far.clone()
+
+            assert len(set(retrieved_so_far.tolist())) == 50
+
+            if num_epochs == 1:
+                with pytest.raises(StopIteration):
+                    next(it)
+
+            if num_epochs in [2, 3, None]:
+                for idx in range(5):
+                    batch = next(it)
+                    # Assert that a new buffer is created inside the loader
+                    assert loader._shuffling_buffer != first_buffer
+                    this_batch = batch['col_0'].clone()
+                    assert list(this_batch.shape)[0] == batch_size
+                    intersection = set(this_batch.tolist()).intersection(set(retrieved_in_first_epoch.tolist()))
+                    assert len(intersection) == batch_size
+                    retrieved_so_far = torch.cat([retrieved_so_far, this_batch], 0)
+
+            if num_epochs == 2:
+                with pytest.raises(StopIteration):
+                    next(it)
+
+            if num_epochs in [3, None]:
+                for idx in range(5):
+                    batch = next(it)
+                    this_batch = batch['col_0'].clone()
+                    assert list(this_batch.shape)[0] == batch_size
+                    retrieved_so_far = torch.cat([retrieved_so_far, this_batch], 0)
+
+            if num_epochs == 3:
+                with pytest.raises(StopIteration):
+                    next(it)
+
+            if num_epochs is None:
+                for idx in range(5):
+                    batch = next(it)
+                    this_batch = batch['col_0'].clone()
+                    assert list(this_batch.shape)[0] == batch_size
+                    retrieved_so_far = torch.cat([retrieved_so_far, this_batch], 0)
+
+
+def test_async_batched_data_loader(two_columns_non_petastorm_dataset):
```

Comment:
Can you please document the test a bit - what's the purpose of the test and what it's testing.

I know `test_batched_data_loader_with_in_memory_cache` is the test that was there before, but if you understand exactly how it works, would appreciate if you could paste that knowledge as comments in the test. I have hard time parsing it :). 

If you haven't dive deep into it, then never mind.

Reply:
sure. Maybe this will be removed after #664 is resolved.